### PR TITLE
Include detailed layer penalties in encumbrance calculations/cache

### PR DESCRIPTION
**Pull request by DoctorVanGogh**
This is a segregated infrastructure piece for my solution to #23425.

It adds/keeps track of the extra layer penalties for armor per bodypart/layer.

It has no _functional_ change on the calculation results, just keeps intermediate calculation values available for later retrieval/display.

### CODE REVIEW

File: src/character.cpp [hunk 1]
```diff
@@ -1488,17 +1491,15 @@ void layer_item( std::array<encumbrance_data, num_bp> &vals,
 
     const int armorenc = !power_armor || !it.is_power_armor() ?
         encumber_val : std::max( 0, encumber_val - 40 );
-
+
```
Comment (Night-Pryanik) [inline src/character.cpp]: Why this four spaces?

File: src/character.cpp [hunk 2]
```diff
@@ -1512,6 +1513,30 @@ bool Character::is_wearing_active_power_armor() const
     return false;
 }
 
+void layer_details::reset() {
+    pieces.clear();
+    max = 0;
+    total = 0;
+}
+
+
```
Comment (Night-Pryanik) [inline src/character.cpp]: Empty line.

File: src/character.cpp [hunk 3]
```diff
@@ -1512,6 +1513,30 @@ bool Character::is_wearing_active_power_armor() const
...
+// The stacking penalty applies by doubling the encumbrance of
+// each item except the highest encumbrance one.
+// So we add them together and then subtract out the highest.
+int layer_details::layer( const int encumbrance ) {
+    pieces.push_back( encumbrance );
+
+    int current = total;
+    if (encumbrance > max) {
```
Comment (Night-Pryanik) [inline src/character.cpp]: Should be `if( encumbrance > max ) {`.

File: src/character.cpp [hunk 4]
```diff
@@ -1512,6 +1513,30 @@ bool Character::is_wearing_active_power_armor() const
...
+        total += max;   // *now* the old max is counted, just ignore the new max
+        max = encumbrance;
+    }
+    else {
```
Comment (Night-Pryanik) [inline src/character.cpp]: Should be `} else {`.

File: src/character.cpp [hunk 5]
```diff
@@ -1529,35 +1554,29 @@ bool Character::is_wearing_active_power_armor() const
...
+void Character::item_encumb( std::array<encumbrance_data, num_bp> &vals,                              
+                             const item &new_item ) const
 {
-    // The highest encumbrance of any one item on the layer.
-    std::array<layer_data, num_bp> layers = {{
-        {{}}, {{}}, {{}}, {{}}, {{}}, {{}}, {{}}, {{}}, {{}}, {{}}, {{}}, {{}}
-    }};
+
+    // reset all layer data
+    for (auto &v : vals)
```
Comment (Night-Pryanik) [inline src/character.cpp]: Should be `for( auto &v : vals )`.

File: src/character.cpp [hunk 6]
```diff
@@ -1529,35 +1554,29 @@ bool Character::is_wearing_active_power_armor() const
...
-    // So we add them together and then subtract out the highest.
+    // make sure values are sane
     for( const body_part bp : all_body_parts ) {
-        for( size_t j = 0; j < MAX_CLOTHING_LAYER; j++ ) {
-            vals[bp].layer_penalty -= std::max( 0, layers[bp][j] );
-        }
-    }
+        encumbrance_data &elem = vals[bp];
+
+		elem.armor_encumbrance = std::max( 0, elem.armor_encumbrance );
```
Comment (Night-Pryanik) [inline src/character.cpp]: Tabs instead of spaces.

File: src/character.cpp [hunk 7]
```diff
@@ -1512,6 +1513,28 @@ bool Character::is_wearing_active_power_armor() const
...
+}
+
+// The stacking penalty applies by doubling the encumbrance of
+// each item except the highest encumbrance one.
+// So we add them together and then subtract out the highest.
+int layer_details::layer( const int encumbrance ) {
+    pieces.push_back( encumbrance );
+
+    int current = total;
+    if ( encumbrance > max ) {
```
Comment (Night-Pryanik) [inline src/character.cpp]: `if` should be glued to the bracket.

File: src/character.cpp [hunk 8]
```diff
@@ -1529,35 +1552,29 @@ bool Character::is_wearing_active_power_armor() const
...
+void Character::item_encumb( std::array<encumbrance_data, num_bp> &vals,                              
+                             const item &new_item ) const
 {
-    // The highest encumbrance of any one item on the layer.
-    std::array<layer_data, num_bp> layers = {{
-        {{}}, {{}}, {{}}, {{}}, {{}}, {{}}, {{}}, {{}}, {{}}, {{}}, {{}}, {{}}
-    }};
+
+    // reset all layer data
+    for ( auto &v : vals )
```
Comment (Night-Pryanik) [inline src/character.cpp]: `for` should be glued to the bracket.

File: src/character.h [hunk 9]
```diff
@@ -61,14 +61,49 @@ enum fatigue_levels {
...
 struct encumbrance_data {
     int encumbrance = 0;
     int armor_encumbrance = 0;
     int layer_penalty = 0;
+
+    std::array<layer_details, static_cast<int>( layer_level::MAX_CLOTHING_LAYER )>
+    layer_penalty_details;
+
+    void layer( const layer_level level, const int emcumbrance ) {
+        layer_penalty += layer_penalty_details[static_cast<size_t>( level )].layer( emcumbrance );
```
Comment (BevapDin) [inline src/character.h]: Why the sudden cast to `size_t`? New code above casts to `int` when the value is used as index or size of an array (note that `std::array` uses `std::size_t` as index/size, not `int`, same for its `operator[]`). At least be consistent.

File: src/character.cpp [hunk 10]
```diff
@@ -1512,6 +1513,28 @@ bool Character::is_wearing_active_power_armor() const
     return false;
 }
 
+void layer_details::reset() {
```
Comment (BevapDin) [inline src/character.cpp]: Seems a simple `*this = layer_details();` would do the same. You can do the same with `encumbrance_data::reset`.

File: src/character.h (hunk 9 above)
Comment (DoctorVanGogh) [reply src/character.h]: You _are_ correct....  This whole thing is a squashed commit, so subsequent lines of code might have been written either at 3am or 5pm. I plead alcohol and/or brainfarts. 

`size_t` it'll be.

File: src/bodypart.h [hunk 11]
```diff
@@ -105,6 +105,7 @@ class body_part_set
         explicit body_part_set( const std::bitset<num_bp> &other ) : parts( other ) { }
 
     public:
+
```
Comment (BevapDin) [inline src/bodypart.h]: Why the empty line?

**Pull request merged**